src/gencost/waterfall.py

```diff
@@ -1708,65 +1841,194 @@ def get_cems_by_x(self, subplant_id_col, xwalk=None, merge_only=False):
-                    subplant_id_col,
+        if subplant_id_col == "generator_id":
+            merged = (
+                self.get_cems()[
+                    [
+                        "plant_id_epa",
+                        "emissions_unit_id_epa",
+                        "report_date",
+                        "generator_starts",
+                        "fuel_starts",
+                        "gross_generation_mwh",
+                        "camd_capacity_mw",
+                    ]
                 ]
-            ]
-            .dropna(axis=0, subset=["plant_id_epa", "emissions_unit_id_epa"])
-            .drop_duplicates(),
-            on=["plant_id_epa", "emissions_unit_id_epa"],
-            how="outer",
-            validate="m:1",
-            indicator=True,
-        )
-        test = (
-            merged.fillna({"plant_id_epa": merged.plant_id_eia})
-            .query("_merge != 'both'")
-            .replace({"_merge": {"left_only": "data_only", "right_only": "xwalk_only"}})
-            .assign(year=lambda x: x.report_date.dt.year)
-            .groupby(["_merge", "year"], dropna=False)
-            .plant_id_epa.nunique()
-            .to_frame()
-            .query("plant_id_epa > 0")
-        )
-        logger.warning(
-            "CEMS %s: Unique plants that did not have matches in both "
-            "CEMS and the xwalk so will be dropped:\n %s \n",
-            {"pf_subplant_id": "prime", "subplant_id": "subplant"}[subplant_id_col],
-            test.squeeze().to_dict(),
-        )
-        if merge_only:
-            return merged
-        aggs = {
-            "generator_starts": "sum",
-            "fuel_starts": "sum",
-            "gross_generation_mwh": "sum",
-            "heat_in_mmbtu": "sum",
-            "co2_tons": "sum",
-        }
-        return (
-            merged.query("_merge == 'both'")
-            .drop(columns=["_merge"])
-            .groupby(["plant_id_eia", subplant_id_col, "report_date"])
-            .agg(aggs | {"camd_capacity_mw": "sum"})
-            .reset_index()
-            .groupby(
-                [
-                    "plant_id_eia",
-                    subplant_id_col,
-                    pd.Grouper(key="report_date", freq="YS"),
-                ],
-                dropna=False,
+                .merge(
+                    xwalk[
+                        [
+                            "plant_id_epa",
+                            "plant_id_eia",
+                            "emissions_unit_id_epa",
+                            subplant_id_col,
+                            "capacity_xwalk",
+                            "subplant_id",
+                        ]
+                    ]
+                    .dropna(axis=0, subset=["plant_id_epa", "emissions_unit_id_epa"])
+                    .drop_duplicates(),
+                    left_on=[
+                        "plant_id_epa",
+                        "emissions_unit_id_epa",
+                        "camd_capacity_mw",
```

Comment:
Do we need to merge on capacity? I hadn't thought of that but could see some need to validate it. My concern would be that, unless `capacity_xwallk` comes from CAMD, it will only rarely match, and merge keys must match exactly.